### DISCUSSION TITLE
sqlalchemy and AWS Aurora DB Connection (Dynamic Access Token)

### DISCUSSION THREAD

**Discussion by ibiswas** [sqlalchemy/sqlalchemy#7818] [Usage Questions]
As AWS IAM Access tokens to access RDS Aurora DB expire after a period of time, how DO I ensure that we fetch the Access token when the existing connections fail because of an expired token/pwd. I have gone through the below link but need a complete example. 
https://docs.sqlalchemy.org/en/14/core/engines.html#engines-dynamic-tokens

I'm currently trying the below code with the event hook of "do.connect" but it doesn't work. Can someone please provide a complete example on how to achieve this.

def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()

def get_password(properties):
    client = boto3.client('rds', region_name=properties['dbRegion'])
    logger.info("Generate db_auth_token")
    password = client.generate_db_auth_token(DBHostname=properties['rdsHost'], Port=properties['rdsPort'],
                                             DBUsername=properties['rdsUsername'])
    logger.info("Generate db_auth_token successfully: " + password)
    return password

SQLALCHEMY_DATABASE_URI = '{}://{}:@{}:{}/{}?ssl_ca={}'.format(config.DB_DIALECT,
                                                               config.DB_RDSUSERNAME,
                                                               config.DB_RDSHOST,
                                                               config.DB_RDSPORT,
                                                               config.DB_RDSSCHEMA,
                                                               config.DB_CERTIFICATE_PATH)

engine = create_engine(  # 2
    SQLALCHEMY_DATABASE_URI,
    # required for sqlite
    connect_args={},  # 3
)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)  # 4
Base = declarative_base()

@event.listens_for(engine, "do_connect")
def receive_do_connect(dialect, conn_rec, cargs, cparams):
    secret = urllib.parse.quote(get_password({'rdsHost': config.DB_RDSHOST,
                                              'rdsPort': config.DB_RDSPORT,
                                              'rdsUsername': config.DB_RDSUSERNAME,
                                              'dbRegion': config.DB_REGION}))
    cparams['password'] = secret

**Comment by zzzeek** [comment]
the recipe refers to how to dynamically generate an access token each time a new connection to the database is created.   it doesn't by itself accommodate an already-connected connection timing out or being cut off due to an expiration of some kind.   If that's the thing which is happening, you also want to make sure that the lifespan of a particular database connection is shorter than that particular expiration time, for which you would use the [pool_recycle](https://docs.sqlalchemy.org/en/14/core/engines.html?highlight=create_engine#sqlalchemy.create_engine.params.pool_recycle) with create_engine().

Otherwise, please provide more specifics about what "doesn't work".  is the event not being called, or is the event called but the connection fails, or does the connection succeed but something goes wrong later, etc.  thanks

**Comment by ibiswas** [comment]
Thanks Mike for a quick turn around on thi , I’ll try the pool recycle
setting, based on the code I provided will the token be fetched during the
the pool recycle though or is it some kind of an internal process?

**Reply by ibiswas** [reply]
Mike also when I introduced the event "@event.listens_for(engine, "do_connect")" as shown in my first post,  I get the following error. I'm assuming once the pool is discarded by the pool_recycle setting I would have to re-fetch the token when a new conn is requested correct. If I remove the event an re-intoduce the token in the URL itself I'm able to connect succesfully.

pymysql.err.OperationalError: (1045, "Access denied for user 'AuroraAppAdmin'@'xx.xxx.xxx.xxx' (using password: YES)")

**Reply by ibiswas** [reply]
So essentially the do_connect does get called but I get the below error. As you can see in the first post all I'm doing is replacing(cparams['password'] = secret) the password in that event after fetching the token.
pymysql.err.OperationalError: (1045, "Access denied for user 'AuroraAppAdmin'@'xx.xxx.xxx.xxx' (using password: YES)")

**Reply by zzzeek** [reply]
1. does the engine connect at all at least once?  or never?  if never, then the routine you are using to get the token is not correct in some way
2. does the engine connect but  then fails later?  then something about the token is going wrong.

the do_connect event runs every time the pool is about to establish a new connection to the database, including when an existing connection is to be "recycled", which isn't actually a "recycle" from the connection's point of view; the connection is closed, and a new one is created (again calling do_connect) to replace it.

**Reply by ibiswas** [reply]
The Engine connects fine when I provide the User Name, Token  without the "do_connect" Event Listener present. Now that I have introduced pool_recycle =300 and the "do_connect" event listener, During the 1st DB query request the "do_connect" event listener is trigerred and immediately throws the below error.
pymysql.err.OperationalError: (1045, "Access denied for user 'AuroraAppAdmin'@'xx.xxx.xxx.xxx' (using password: YES)")

**Reply by ibiswas** [reply]
So this works
####################
from sqlalchemy import create_engine, event
from sqlalchemy.orm import sessionmaker, Session
from sqlalchemy.ext.declarative import declarative_base
from app import config
import urllib
import boto3, json
import logging

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)
fmt_string = '%(asctime)s, %(levelname)-8s [%(filename)s:(args):%(lineno)d]:%(message)s'
logging.basicConfig(level=logging.INFO, format=fmt_string, datefmt='%d-%m-%Y:%H:%M:%S')


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


def get_password(properties):
    client = boto3.client('rds', region_name=properties['dbRegion'])
    logger.info("Generate db_auth_token")
    password = client.generate_db_auth_token(DBHostname=properties['rdsHost'], Port=properties['rdsPort'],
                                             DBUsername=properties['rdsUsername'])
    logger.info("Generate db_auth_token successfully: " + password)
    return password


SQLALCHEMY_DATABASE_URI = '{}://{}:{}@{}:{}/{}?ssl_ca={}'.format(config.DB_DIALECT,
                                                                 config.DB_RDSUSERNAME,
                                                                 urllib.parse.quote(
                                                                     get_password({'rdsHost': config.DB_RDSHOST,
                                                                                   'rdsPort': config.DB_RDSPORT,
                                                                                   'rdsUsername': config.DB_RDSUSERNAME,
                                                                                   'dbRegion': config.DB_REGION})),
                                                                 config.DB_RDSHOST,
                                                                 config.DB_RDSPORT,
                                                                 config.DB_RDSSCHEMA,
                                                                 config.DB_CERTIFICATE_PATH)

engine = create_engine(  # 2
    SQLALCHEMY_DATABASE_URI, pool_recycle=300, pool_pre_ping=True
)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)  # 4
Base = declarative_base()
#####################

The below doesn't work
####################
from sqlalchemy import create_engine, event
from sqlalchemy.orm import sessionmaker, Session
from sqlalchemy.ext.declarative import declarative_base
from app import config
import urllib
import boto3, json
import logging
from cachetools import TTLCache, cached

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)
fmt_string = '%(asctime)s, %(levelname)-8s [%(filename)s:(args):%(lineno)d]:%(message)s'
logging.basicConfig(level=logging.INFO, format=fmt_string, datefmt='%d-%m-%Y:%H:%M:%S')


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


def get_password(properties):
    client = boto3.client('rds', region_name=properties['dbRegion'])
    logger.info("Generate db_auth_token")
    password = client.generate_db_auth_token(DBHostname=properties['rdsHost'], Port=properties['rdsPort'],
                                             DBUsername=properties['rdsUsername'])
    logger.info("Generate db_auth_token successfully: " + password)
    return password


SQLALCHEMY_DATABASE_URI = '{}://@{}:{}/{}?ssl_ca={}'.format(config.DB_DIALECT,
                                                                 config.DB_RDSHOST,
                                                                 config.DB_RDSPORT,
                                                                 config.DB_RDSSCHEMA,
                                                                 config.DB_CERTIFICATE_PATH)

engine = create_engine(  # 2
    SQLALCHEMY_DATABASE_URI, pool_recycle=300, pool_pre_ping=True, echo=True
)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)  # 4
Base = declarative_base()


@event.listens_for(engine, "do_connect")
def receive_do_connect(dialect, conn_rec, cargs, cparams):
    secret = urllib.parse.quote(get_password({'rdsHost': config.DB_RDSHOST,
                                              'rdsPort': config.DB_RDSPORT,
                                              'rdsUsername': config.DB_RDSUSERNAME,
                                              'dbRegion': config.DB_REGION}))
    cparams['password'] = secret
    cparams['user'] = config.DB_RDSUSERNAME


###################

**Reply by zzzeek** [reply]
OK that's not a connection.  when you make an engine, it just takes in the arguments, nothing happens.   

make the engine using create_engine(), then say engine.connect().   I think that will fail here.  so you have yet to actually connect to the DB.

**Reply by ibiswas** [reply]
The Usage of the DB Code is in a different file, Here is an example


@my_router.get(path="/id}", response_model=SomeSchema,
               description="Some Desc")
def get(id: str, db: Session = Depends(get_db)):
    try:
        data= db.query(ModelObj).filter(
            ModelObj.id == id).first()
    except HTTPException as he:
        raise he
    except Exception as e:
        raise HTTPException("Something wrong happened", 500)

**Reply by ibiswas** [reply]
Thanks Mike i was able to resolve, my issue was the encoding being done on top of the token generation being done. I just had to remove that and the raw token worked, So I had to remove the  urllib.parse.quote
